src/game/counting.rs

```diff
@@ -0,0 +1,276 @@
+use crate::game::cards::Card;
+use crate::game::scoring::{score_run, Combination, CombinationKind, Score};
+use std::error::Error;
+
+/// Calculates the score during the counting phase where played_cards have already been played and card is now played
+///
+///
+pub fn score_counting_cards_played(
+    played_cards: &[Card],
+    card: Card,
+) -> Result<Score, Box<dyn Error>> {
+    let count: i32 = played_cards.iter().map(|c| c.value).sum::<i32>() + card.value;
+    if count > 31 {
+        return Err("invalid card + count > 31".into());
+    }
+
+    let mut score: Score = Score::new();
+    let mut all_cards: Vec<Card> = played_cards.to_vec();
+
+    all_cards.push(card);
+    if count == 15 {
+        score.combinations.push(Combination::new(
+            CombinationKind::Fifteen,
+            all_cards.clone(),
+        ));
+    } else if count == 31 {
+        score.combinations.push(Combination::new(
+            CombinationKind::ThirtyOne,
+            all_cards.clone(),
+        ));
+    }
+
+    //
+    //  check to see if we have matching rank
+    let mut len = played_cards.len();
+    let mut count = 0;
+    if len > 0 {
```

Comment:
Can you just reverse iterate and count without the extra guards? If there are only 2-4 cards total and they're they same rank, you'll break the loop with a count. If the card doesn't match, you'll break before you get to the end. I feel like you could cut two `if`s with something like that.

played_cards.iter().rev() { ... }